[MRG] Invariance test for non-scalar metric

dmohns opened the pull request:
#### Reference Issues/PRs

Following up a discussion from https://github.com/scikit-learn/scikit-learn/pull/10591

Automated invariance test are currently implemented in `test_common.py` for scalar ranking functions. It is desirable to have these tests also be available for ranking metrics with potentially non-scalar outputs. For example `roc_curve`, `classification_report` or `confusion_matrix`.


#### What does this implement/fix? Explain your changes.

Scope of this PR is to generalize `test_common.py` to allow testing of metrics with non-scalar output. This shall be achieved by replacing all assert in such a away that they allow assertion on array-like structures. 

#### Any other comments?

It should first be discussed if a solution like this is feasible. An alternated route suggested by @jnothman is to aggregate non-scalar metrics using an arbitrary measure and apply the scalar invariance tests.

#### To-do
- [x] Add other metrics to the tests
- [ ] Add wrappers for `assert_array_greater` and `assert_array_not_equal`
- [x] Check if certain tests become obsolete (if so remove them)

dmohns committed:
Support non-scalar output in metrics invariance tests

dmohns committed:
Add curves to invariance tests

jnothman commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -951,33 +974,36 @@ def check_sample_weight_invariance(name, metric, y1, y2):
-            "equal (%f) for %s" % (weighted_score, name))
+    assert_raises(
+        AssertionError,
+        assert_array_equal, unweighted_score, weighted_score,
```
Use approximate equality here, please

dmohns replied:
You mean `assert_array_almost_equal` right? If I do that it will make the test fail `micro_average_precision_score`

```
    def _raiseFailure(self, standardMsg):
        msg = self.test_case._formatMessage(self.msg, standardMsg)
>       raise self.test_case.failureException(msg)
E       AssertionError: Unweighted and weighted scores are unexpectedly almost equal (0.118667604216) and (0.118668873707) for micro_average_precision_score

msg        = 'Unweighted and weighted scores are unexpectedly almost equal (0.118667604216) and (0.118668873707) for micro_average_precision_score'
self       = <sklearn.utils._unittest_backport._AssertRaisesContext instance at 0x7ff47bf8f758>
standardMsg = 'AssertionError not raised'
```

jnothman replied:
I think micro_average_precision_score should be identical to accuracy_score. Perhaps this is not as robust a test as we thought?  Does that still fail if assert_allclose is used? It relies on relative tolerance rather than absolute and is probably a better test here

dmohns replied:
Interesting. The test passes when using `assert_allclose`. 

Additionally the use of `assert_array_almost_equal` is dis-advised by NumPy. I decided to replace all approximate equality checks with `assert_allclose` to increase overall robustness of the invariance tests.

jnothman reviewed (approved):
Who knew it could be this easy! Nice

dmohns committed:
Apply invariance tests to curve metrics

dmohns committed:
Test sample weight invariance using approximation

- Increase weight range to circumvent approximate equality for some
metrics.

dmohns commented:
I am not 100% sure if we would need wrappers for the array asserts `greater` and `not_equal`?  Especially the non-equality test is a bit cumbersome. To the unfamiliar reader this might disguise what the test is actually trying to check.
On the other hand these checks are only ever used in `test_common.py`. Adding abstract versions of the above mentioned asserts would add unneeded complexity without benefits to other parts of scikit-learn.

What do you think?

PS: I gave up on trying to add `classification_report` to `test_common.py`. The fact that it returns a string makes it sheer impossible to write a reasonable assertion.

dmohns committed:
Remove obsolete metrics tests

dmohns committed:
Test confusion matrix with scaling

dmohns committed:
Make invariance tests more robust

- The use of `assert_allclose` over `assert_array_almost_equal` is
recommended by NumPy.
- Revert initial change of sample weights as obsolete.

dmohns committed:
Sort imports alphabetically

jnothman reviewed (approved):
I'm not certain I've reviewd this thoroughly, but it LGTM

jnothman reviewed (approved):
I'm not certain I've reviewd this thoroughly, but it LGTM

amueller commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -103,7 +96,13 @@
     "unnormalized_accuracy_score": partial(accuracy_score, normalize=False),
-    "confusion_matrix": confusion_matrix,
+
+    "unnormalized_confusion_matrix": confusion_matrix,
```
I think it's a bit weird to name them this way, given that ``confusion_matrix`` is unnormalized for us. I'd rather call this one ``confusion_matrix`` and the other one ``normalized_confusion_matrix``.

dmohns replied:
Agreed. It felt very counter-intuitive to name `confusion_matrix` that way.
However, this was necessary to have the scaling test schedule the correct metric.
See https://github.com/dmohns/scikit-learn/blob/94b9bd0d52ee55478ea7607cfd3037c83e4c2f4b/sklearn/metrics/tests/test_common.py#L1009

For the scope of this PR I didn't want to touch any of the actual test code. 

The other option would be to introduce a new list of "unnormalized" metrics and 
use that for scheduling rather than `name.startswith`.

Thoughts?

dmohns replied:
Basically the ones that can be called with a `normalize` which defaults to `True` parameter. Currently this affects `balanced_accuracy_score`, `jaccard_similarity_score`, `zero_one_loss` and `log_loss`. Let me add a comment on why I am doing what I am doing there and see if it gets clearer.

amueller commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -389,27 +403,29 @@ def test_symmetry():
         set(ALL_METRICS))
 
-    assert_equal(
+    assert_array_equal(
```
I don't understand this one

dmohns replied:
Good catch. This is an artifact of me horse-blindly removing all occurrences of
`assert_equal` where technically possible. In this situation, it is more clear to keep to original `assert_equal` to avoid confusion.

Will undo this change.

amueller commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -148,6 +147,14 @@
+# Metrics are tested with numpy's assert_array. Therefor box-shaped outputs
+# are required.
+CURVE_METRICS = {
+    "roc_curve": lambda *args, **kwargs: roc_curve(*args, **kwargs)[:2],
```
you're throwing away the thresholds, right? Which makes sense, but I think the comment should say that. That's also not only about the shape right? The thresholds probably don't behave like a metric.

dmohns replied:
Yes, I am throwing away the thresholds here. But no, I am only really doing it because of shape, as `thresholds` will have a different dimension compared to other returned values.
  
If I recall correctly the `thresholds` on their own should pass all the tests. I will test if we can keep the `thresholds` by padding the result into a box-shaped array rather than throwing stuff away.

amueller reviewed (commented):
Looks good, minor comments.

dmohns committed:
Revert to assert_equal for list asserts

In this situation, it is more clear to keep to original assert_equal to
avoid confusion.

dmohns committed:
Clarify unnormalized behaviour of confusion_matrix

dmohns committed:
Add alternative way to handle precision_recall_curve

jnothman commented:
@dmohns, could you please resolve conflicts with master. We'll try to merge soon, given that @amueller and I have both more-or-less approved.

dmohns commented:
@jnothman Sure, will do.

dmohns committed:
Merge branch 'master' into invariance-test-non-scalar-metric

Resolved Merge Conflicts:
- sklearn/metrics/tests/test_common.py

jnothman commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -148,6 +152,36 @@
 }
 
+
+def precision_recall_curve_padded_thresholds(*args, **kwargs):
```
To be sure, we could pad with something more conspicuous, like nan???

Otherwise lgtm

dmohns replied:
I like the idea of padding with NaNs. It makes things a bit more verbose.

dmohns committed:
Make padding more conspicuous

Commit:
Make metrics/test_common run with old NumPy versions

- Specifically NumPy 1.10.4 seems to have a problem with casting 
```
((val1, val2), )                  => ((val1, val2), (val1, val2), ...)
```
see https://github.com/numpy/numpy/blob/v1.10.4/numpy/lib/arraypad.py#L989

We can circumvent this by using the simpler trait of np.pad, passing a singleton to `constant_values`.

Commit:
Make metrics/test_common run with ancient NumPy versions

- Specifically NumPy 1.8.2 seems to have a problem with casting 
```
int                               => ((int, int), (int, int), ...)
```
when shape is NaN. See https://github.com/numpy/numpy/blob/v1.8.2/numpy/lib/arraypad.py#L1042

We can circumvent this by wrapping NaN value into list instead of passing a NaN-singleton to `constant_values`.

jnothman commented:
Thanks! Good work!
